src/Workspaces/CoreTestUtilities/MEF/UseExportProviderAttribute.cs

```diff
@@ -91,10 +87,34 @@ public override void Before(MethodInfo methodUnderTest)
         /// </remarks>
         public override void After(MethodInfo methodUnderTest)
         {
-            var exportProvider = ExportProviderCache.ExportProviderForCleanup;
             try
             {
-                if (exportProvider?.GetExportedValues<IAsynchronousOperationListenerProvider>().SingleOrDefault() is { } listenerProvider)
+                DisposeExportProvider(ExportProviderCache.LocalExportProviderForCleanup);
+                DisposeExportProvider(ExportProviderCache.RemoteExportProviderForCleanup);
+            }
+            finally
+            {
+                // Replace hooks with ones that always throw exceptions. These hooks detect cases where code executing
+                // after the end of a test attempts to create an ExportProvider.
+                MefHostServices.TestAccessor.HookServiceCreation(DenyMefHostServicesCreationBetweenTests);
+                RoslynServices.TestAccessor.HookHostServices(() => throw new InvalidOperationException("Cannot create host services after test tear down."));
+
+                // Reset static state variables.
+                _hostServices = null;
+                ExportProviderCache.SetEnabled_OnlyUseExportProviderAttributeCanCall(false);
+            }
+        }
+
+        private static void DisposeExportProvider(ExportProvider? exportProvider)
+        {
+            if (exportProvider == null)
```

Comment:
💡 Due to the way this method is now structured, you can use this form and remove the `try`/`finally`:

```csharp
using var _ = exportProvider;
```